server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java

```diff
@@ -578,7 +593,12 @@ static ReducedQueryPhase reducedQueryPhase(
         final SearchProfileResultsBuilder profileBuilder = profileShardResults.isEmpty()
             ? null
             : new SearchProfileResultsBuilder(profileShardResults);
-        final SortedTopDocs sortedTopDocs = sortDocs(isScrollRequest, bufferedTopDocs, from, size, reducedCompletionSuggestions);
+        final SortedTopDocs sortedTopDocs = rankContext == null
+            ? sortDocs(isScrollRequest, bufferedTopDocs, from, size, reducedCompletionSuggestions)
+            : rankContext.rank(queryResults.stream().map(SearchPhaseResult::queryResult).toList(), topDocsStats);
```

Comment:
The `ScoreDoc`s as part of `SortedTopDocs` in this case should be instances of `RankResult`. Given the `RankContext` is the only thing that is currently able to process rank information, I thought it made the most sense to do a cast with the `decorateSearchHit` method and possibly add other methods to the `RankContext` if we need this information elsewhere. There is already a precedent for this type of design when handling sort because we check for instanceof `FieldDoc` and cast as well.

I'm definitely open to changing this, but I don't have a good feel for either what should become part of `SortedTopDocs` and where else it would be used.

Reply:
I think for the first iteration this should continue to stay as it is because it requires no modifications to the fetch phase at this time, and we are already modifying a lot of the search phases code.